Support mapping volumes in Dev Service container based databases

Sgitario opened the pull request:
This is a very useful feature that allows mapping volumes in the Dev Service containers. 

For example: to keep the Postgres data folder in the localsystem and hence have the data persistently. 

Fix https://github.com/quarkusio/quarkus/issues/30595

geoand commented:
I think this is a great feature!

But I'll defer reviewing to @yrodiere since it involves the Datasource code

yrodiere reviewed (commented):
Thanks.

The feature looks nice indeed, but there's some danger involved as Hibernate ORM will clear the database on startup by default. So if you mount a volume on your devservice, and don't take care to configure Hibernate ORM appropriately, you will lose data.

Granted, I don't expect any sensible person to mount valuable data there, but still, I think some more documentation is in order:

1. Clarify in the javadoc of that new configuration property that data contained in mounted volumes may be altered by the application in dev mode (and tests!), and may be completely wiped out.
2. Maybe add a subsection with  an example in the databases-dev-services documentation, explaining in more details the risks (in particular with Hibernate ORM, but probably also Flyway, ...)? For ORM, essentially [if you don't use Flyway/Liquibase and don't set `quarkus.hibernate-orm.database.generation` your devservice schema and data will be wiped out on startup](https://github.com/quarkusio/quarkus/blob/0a503ffb78194d083552fea030828eaf3976998d/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L282-L301); if you use Flyway/Liquibase or set `quarkus.hibernate-orm.database.generation` to `none` or `validate` you're safe.

stuartwdouglas commented:
The code LGTM, but it would be great to see some docs with real world (e.g. postgres and mysql) examples of how you actually do this. I know you could figure it out by reading the relevant containers docs but a copy and paste example would be way more user friendly imho.

Sgitario commented:
Thanks for your suggestions! I will be working on this.

Sgitario commented:
PR updated with more documentation and a couple of examples. I verified each of these examples using the following tests:

```java
public class DevServicesPostgresqlDatasourceWithFileVolumeTestCase {

    @RegisterExtension
    static QuarkusUnitTest test = new QuarkusUnitTest()
            .overrideConfigKey("quarkus.datasource.db-kind", "postgresql")
            .overrideConfigKey("quarkus.datasource.devservices.volumes.\"/home/jcarvaja/sources/tmp/data\"",
                    "/var/lib/postgresql/data");

    @Inject
    DataSource ds;

    // I executed this test to create my database schema and the data, and then I disabled to only execute the following one.
    @Disabled 
    @Test
    public void create() throws Exception {
        try (Connection con = ds.getConnection()) {
            con.createStatement()
                    .execute("CREATE TABLE PERSON (name VARCHAR(50) PRIMARY KEY)");
            con.createStatement()
                    .execute("INSERT INTO PERSON (name) VALUES ('JOHN')");
        }
    }

    // In the second round, this test passed!
    @Test
    public void testDatasource() throws Exception {
        String name = "";
        try (Connection con = ds.getConnection();
                CallableStatement cs = con.prepareCall("SELECT name FROM PERSON");
                ResultSet rs = cs.executeQuery()) {
            if (rs.next()) {
                name = rs.getString(1);
            }
        }
        assertEquals("JOHN", name, "The volume was not properly mounted");
    }
}
```

And for MySQL:

```java
public class DevServicesMySqlDatasourceWithFileVolumeTestCase {

    @RegisterExtension
    static QuarkusUnitTest test = new QuarkusUnitTest()
            .overrideConfigKey("quarkus.datasource.db-kind", "mysql")
            .overrideConfigKey("quarkus.datasource.devservices.volumes.\"/home/jcarvaja/sources/tmp/mysqldata\"",
                    "/var/lib/mysql");

    @Inject
    DataSource ds;

    // I executed this test to create my database schema and the data, and then I disabled to only execute the following one.
    @Disabled 
    @Test
    public void create() throws Exception {
        try (Connection con = ds.getConnection()) {
            con.createStatement()
                    .execute("CREATE TABLE PERSON (name VARCHAR(50) PRIMARY KEY)");
            con.createStatement()
                    .execute("INSERT INTO PERSON (name) VALUES ('JOHN')");
        }
    }

    // In the second round, this test passed!
    @Test
    public void testDatasource() throws Exception {
        String name = "";
        try (Connection con = ds.getConnection();
                CallableStatement cs = con.prepareCall("SELECT name FROM PERSON");
                ResultSet rs = cs.executeQuery()) {
            if (rs.next()) {
                name = rs.getString(1);
            }
        }
        assertEquals("JOHN", name, "The volume was not properly mounted");
    }
}
```

Note that to make Postgresql worked, I had to modify the log waiting logic because of one issue in Testcontainers side.

yrodiere commented on docs/src/main/asciidoc/databases-dev-services.adoc:
```diff
@@ -59,6 +59,55 @@ ibmcom/db2:11.5.0.0a
 
+== Mapping volumes into Dev Services for Database
+
+Mapping volumes from our local system to the containers is handy to provide files from local to the database like scripts or configuration, but also to keep the whole database data that can be reused by a later execution.
```
```suggestion
Mapping volumes from the Docker host's filesystem to the containers is handy to provide files like scripts or configuration, but also to preserve database data and reuse it after an application restart.
```

yrodiere commented on docs/src/main/asciidoc/databases-dev-services.adoc:
```diff
@@ -59,6 +59,55 @@ ibmcom/db2:11.5.0.0a
+
+[NOTE]
+====
+Mapping volumes will only work in Dev Services with a container based database like PostgreSQL. 
```
```suggestion
Mapping volumes will only work in Dev Services with a container-based database like PostgreSQL. 
```

yrodiere commented on docs/src/main/asciidoc/databases-dev-services.adoc:
```diff
@@ -59,6 +59,55 @@ ibmcom/db2:11.5.0.0a
+Mapping volumes will only work in Dev Services with a container based database like PostgreSQL. 
+====
+
+Dev Services supports volumes from the filesystem and the classpath:
```
```suggestion
Dev Services volumes can be mapped to the filesystem or the classpath:
```

yrodiere commented on docs/src/main/asciidoc/databases-dev-services.adoc:
```diff
@@ -59,6 +59,55 @@ ibmcom/db2:11.5.0.0a
+quarkus.datasource.devservices.volumes."/local/test/data"=/var/lib/mysql
+----
+
+When starting the dev services (for example, in tests or in DEV mode), you will see that the folder "/local/test/data" will be created at your file sytem and that will contain all the database data. When rerunning again the same dev services, this data will contain all the data you might have created beforehand. 
```
```suggestion
When starting Dev Services (for example, in tests or in DEV mode), you will see that the folder "/local/test/data" will be created at your file sytem and that will contain all the database data. When rerunning again the same dev services, this data will contain all the data you might have created beforehand. 
```

yrodiere commented on docs/src/main/asciidoc/databases-dev-services.adoc:
```diff
@@ -59,6 +59,55 @@ ibmcom/db2:11.5.0.0a
+
+[IMPORTANT]
+====
+When mapping volumes from your local file system to keep the database data, Quarkus might wipe out all the data if you have configured your application to drop and create the schema using `quarkus.hibernate-orm.database.generation = drop-and-create` (using `quarkus.hibernate-orm.database.generation=none` or `quarkus.hibernate-orm.database.generation=validate` is recommended). Also, using Flyway t ...
```
No, the data is at risk even if you didn't configure anything. That's the thing: for dev services, we default to wiping out the data...

```suggestion
When using Dev Services with Hibernate ORM, by default Quarkus will wipe out the database on application startup, which will wipe out the database data on your Docker host's filesystem.
Configure `quarkus.hibernate-orm.database.generation=none` or `quarkus.hibernate-orm.database.generation=validate` to avoid this behavior.

Also, using Flyway to migrate your schema when starting the application will modify the database data on your Docker hosts's file system.
```

yrodiere commented on extensions/datasource/runtime/src/main/java/io/quarkus/datasource/runtime/DevServicesBuildTimeConfig.java:
```diff
@@ -80,8 +80,21 @@ public class DevServicesBuildTimeConfig {
+     * When using a file system location, the volume will be created with read-write permission, so the data in your file
+     * system might be wiped out or altered.
+     *
+     * If the provider is not container based (e.g. an H2 or Derby Database) then this has no effect.
```
Maybe add some code to throw an exception if that happens?

Sgitario replied:
There are other properties that are container-based and that does not have this exception, so I would not over-complicate the logic here. We can always add it if it's really necessary.

yrodiere commented on extensions/jdbc/jdbc-postgresql/deployment/src/test/java/io/quarkus/jdbc/postgresql/deployment/DevServicesPostgresqlDatasourceWithVolumeTestCase.java:
```diff
@@ -0,0 +1,44 @@
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class DevServicesPostgresqlDatasourceWithVolumeTestCase {
```
Note you could probably take advantage of `QuarkusDevModeTest` to test that data is persisted upon restart... Though I don't know how to force a restart, except by modifying a file like in `io.quarkus.hibernate.orm.HibernateHotReloadTestCase#testImportSqlWithContinuousTesting`.

Sgitario replied:
This test is not mapping the db data, just using a volume where there is a script. 
Maybe we could add another test case to verify this use case using QuarkusDevModeTest (actually, I have these tests in another repository because I wanted to be sure about it), but I don't think it's really necessary to be covered here.

stuartwdouglas replied:
The container is not restarted on hot reload, so testing by modifying the file does not really prove anything with regards to how container volumes are handled.

yrodiere replied:
Right. We'd need to modify `application.properties` and change very specific configuration properties to trigger a restart:

https://github.com/quarkusio/quarkus/blob/5e78877a5f0c6592e9663824b8536a4f44288904/extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java#L75-L101

Sgitario replied:
I think this is out of the scope of this pull request.

yrodiere replied:
I think testing the features we document in a PR ("is handy [...] to preserve database data and reuse it after an application restart") is very much in scope of that PR.
But I'll grant you this opinion isn't widely shared.

yrodiere reviewed (commented):
Thanks! A few suggestions below, mostly about wording.

Sgitario committed:
Support mapping volumes in Dev Service container based databases

This is a very useful feature that allows mapping volumes in the Dev Service containers. 

For example: to keep the Postgres data folder in the localsystem and hence have the data persistently. 

Fix https://github.com/quarkusio/quarkus/issues/30595

Co-authored-by: Yoann Rodière <yoann@hibernate.org>

Sgitario commented:
I think the failures are unrelated to these changes.
Would you mind having another round to review this PR? @yrodiere @stuartwdouglas

yrodiere reviewed (approved):
I think the feature is ready, and it's nice. Thanks for that.

I also think the tests you ran manually (and posted somewhere in a comment) should be automated and included in this PR, because otherwise they'll never make it into the code base and then we won't notice regressions.

That being said, feel free to merge if you don't want to work on those tests.

Sgitario commented:
> I think the feature is ready, and it's nice. Thanks for that.
> 
> I also think the tests you ran manually (and posted somewhere in a comment) should be automated and included in this PR, because otherwise they'll never make it into the code base and then we won't notice regressions.
> 
> That being said, feel free to merge if you don't want to work on those tests.

I feel that the example only exposes a very handy use case and yes, ideally it should be covered by tests, but taking into account the complexity, let's merge it as it is and we can add tests if required in the future. 

The test failures are unrelated. Merging
